### ISSUE TITLE
Feature request - Content localizable

### ISSUE THREAD

**Issue by AriasBros** [BookStackApp/BookStack#569]
### Desired Feature

In my organization we need to have our content in several languages. 
Do you have in your roadmap implement localizable content?

##### Current Behavior

In this moment we can create books/chapters/pages in the languages we need, but translations have no connection to each other and all content in all languages appears to the user.

##### Expected Behavior

Would be fantastic if only the content in the current language of the user is shown, with links to the available translations and fallback to the default language if missing translation.

If you have not planned this functionality yet, I would like to collaborate with you by implementing it.

**Comment by ssddanbrown** [comment]
Hi @AriasBros,
Thank you for your suggestion. This is currently not on the roadmap as it has not been requested before.

To be totally honest this is quite a chunky feature that will require maintenance so I'd be apprehensive about including this without a very high demand.

**Comment by AlexWerz** [comment]
I plan to use BookStack to add books in multiple languages. The user can then decide which language of the book (in my case manuals) he wants to read. This makes updates easier as they often just need to be translated from the english master.
As this is not a feature (yet?) I plan on using tags to set the language and book shelfs to organize languages. Are there any other approaches currently used for maintaining different languages of the same book?

**Comment by Drakone** [comment]
COmmunecter aimerait également pouvoir proposer sa documentation en plusieurs langues : doc.co.tools.

**Comment by Orchal** [comment]
Hi,

I'm also very interested by this feature as I work in a public research organization which is international.
Sure we can use tags and double all pages but it would be very nice to have a language switch somewhere and only have one language displayed at the same time.

By the way thank you for all your work, this is pretty good!

**Comment by lreiher** [comment]
I also want to express my need for a multi-language wiki. Of course most of us understand English nowadays, but it would definitely help a lot if certain pages could simply be read in your preferred language with English as the fallback.

Thanks for all your work so far!

**Comment by TobiasDax** [comment]
Any updates on this? Is there a roadmap for this feature?

**Comment by ssddanbrown** [comment]
@TobiasDax I'm afraid not, no updates and this is currently not on the roadmap.

**Comment by pabst2k** [comment]
First of all: Thanks for this amazing software! I really love it!

I also have a need for a translation feature. I write manuals with BookStack and would make my life so much easier. 
You said it will be hard to implement. Maybe starting with a "lite" approach would make it easier. 
What I mean is e.g.:
1. Click on edit page 
1.1 Click on create translation
2. Page is copied to \language\pageName
3. Translation can be done by just replacing the old text line by line more or less 
4. Save

I know it will be hard to deal with updates, this would require more complex code. But with this solution at least the pages are directly connected to each other and I can switch between them. And other editor will instantly see that there is another translation that needs to be updated. In the bookshelf he will forget about that.

**Comment by Drakone** [comment]
Here is the method I used while waiting for the new feature: : https://doc.co.tools/shelves

**Comment by pabst2k** [comment]
@Drakone so you basically have two shelves for different languages or did I miss something? As I wrote having multiple shelves / books is not really what I want because it makes it hard to connect the single pages directly with each other.

**Comment by tgraveleau** [comment]
Hey,
We have customers from different country and we would appreciate this feature too.
Hope it will be available.

**Comment by DiscoveryOV** [comment]
Our organization has a need for this. Throwing in my hat for support.

**Comment by ferranrego** [comment]
Hello! I think this could be an amazing feature to have, as Bookstack could work also as a Help Center. I would love to see it for my project, any update on this? Did you reconsider adding this feature on the roadmap?

Thanks!

**Comment by Ombrelin** [comment]
Hello, I love Bookstack and I second what was said : translatable books would be a great new feature

**Comment by gagarine** [comment]
We are publishing free school book. We are using automatic translation to make them accessible to a wider audience. We are looking to a new platform but content translation is a must.

**Comment by digitall-it** [comment]
You guys are incredible. Thank you.

**Comment by dhigby** [comment]
We live in a multilingual world.  I love the BookStack app and experimented with the demo site.  It is a great solution, one of the best I've seen, but I can't go with it because my Book/manual needs to be kept up to date in between 2-7 different languages.

**Comment by ssddanbrown** [comment]
Just to re-confirm my position on this, since it's been a while:

I appreciate this would help out a bunch of different use-cases, but it would be a fairly bulky feature to implement while being hard & time consuming to work this into the platform in a manner that does not hinder the user-experience of existing non-multi-linugal use-cases. Additionally, the kind of changes this would entail would have an ongoing dragging cost on future further development. 

I was recently approached by a charity that wanted to use BookStack, with this feature being a requirement; and while I really wanted to support them by looking to implement this feature, I could not warrant the cost it would have, at least at this point in the project's life.

**Comment by SubJunk** [comment]
I would be very interested in this as well. If I contributed a Pull Request to make a version of the feature, would you be interested, or still no?

**Comment by ssddanbrown** [comment]
@SubJunk thanks for offering but likely not due to my concerns above.

**Comment by soflane** [comment]
First, thanks so much for this amazing soft! I'm starting using it as much for my work and personal life. I came from wiki.js and it's way better for my usage, and for an internal doc use In a business. But indeed I had translation in wikijs and it's kinda a pain in the ass where I have different books with plenty of pages to manage. I would looove to have this feature implemented, as I already love this app and I will use it everyday more!
Big up for this soft!
Peace, love & unity ✌️

**Comment by AliShaatani** [comment]
thank you for this awesome piece of application, 
but, as you can see, you can't use it, as an official knowledge base section, without multi-language selection 
since the public visitor may want to change language, in certain places like technical documentation, where you need to use universal language, which is English in this case, and other languages in places like help and how to

**Comment by ssoulaimana** [comment]
+1 for multi language support in Bookstack.
Thanks

**Comment by jasell** [comment]
+1 for multi-language support and easy/predefined integration to translation management system (like Crowdin or other).

**Comment by robertfshort** [comment]
+1 for this feature

**Comment by ffranchina** [comment]
Hello everyone! I was about to open a feature request for a kind of relatable feature but I thought that maybe would be better to ask here first and, in case, discuss it.

**Proposal**
From the perspective of the semantic web it would be quite nice to be able to specify the language of a single book though the attribute `lang` (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang): it can improve the indexing and the accessibility of the resources and it can be useful also for retrieving the resources in an automatable way.
Such change should be quite easy to implement since it is just an additional attribute to store for every book and could be shown only at html-level (for elements that are `data-entity-type="book"`, for instance). 

Such feature could also be implemented in a less invasive way though the usage of the already implemented tags but it would require anyway some custom logic in order to place the respective `lang` attribute in the html.

**Extensions**
Having such feature implemented could also open to some interesting scenarios:
 - By means of some CSS would be possible to show only items written in a certain language
 - Could allow to write some bots to retrieve only resources written in a specific language
 - Up to some extent, a kind of multi-language experience can be achieved with few customization of the code

**Conclusions**
I think this would be a valuable addition by itself, but can also be seen as starting point for interesting scenarios. I think I will end up implementing such feature by myself but, if welcomed by @ssddanbrown, I would be very happy of implementing it according to his view of the project!

Thanks!

**Comment by ssddanbrown** [comment]
Hi @ffranchina, Thanks for putting forward the idea. From my perspective I'd rather crawl towards a fuller solution than sooner implement a half-solution that does not have direct and/or clear usage/benefit within the system itself. I can see how your proposed idea may help in some scenarios but I try not to implement features based upon possibilities or opening interesting scenarios, Instead I try to focus on fundamental desires that fit the usage of BookStack and the vocal desire is proper multi language page support.

If your proposal would help you at the current time, I believe this could be done with tags (as you mention) along with the [visual theme system](https://github.com/BookStackApp/BookStack/blob/development/dev/docs/visual-theme-system.md) to apply the `lang` tags where required. If you'd like some further guidance on that feel free to open a new issue.

---

Just to state my updated view on this, since it's been almost a year, I'm still aligned with [my previous view](https://github.com/BookStackApp/BookStack/issues/569#issuecomment-788335561) but there's a potential chance I may soon transition to work on the project full time and hence it may be more feasible to include. Either way things go, page-translations will be in mind when making other fundamental changes to slowly get us moving towards the feature, or at least not fully block us off.

**Comment by jasell** [comment]
What are your plans for managing the translations? Do you plan to integrate with existing TMS systems, like Crowdin? Or are you planning to build your own translation support?
I spoke to Crowdin and they are happy to support as they wish to see more platforms integrating their support (as they gain business from that).

**Comment by ssddanbrown** [comment]
> What are your plans for managing the translations? Do you plan to integrate with existing TMS systems, like Crowdin? 

I have no specific plans for that. The primary goal would be multiple languages per page in the current editor experiences. Automation outside of that would probably be delegated to API use via external apps/middleware. Doesn't make sense for us to support specific translation services/platforms.

**Comment by jasell** [comment]
I would agree that every page would need its separate "copy" in other languages. I would not publish the translated version nested with the original language. I would make parallel books, or bookshelves in your case (i'm not sure what make most sense). As a user would not jump between language versions when consuming the book. They typically have a preset for what language they wish to use.

So when storing each book I would use a folder for each language high up  in the hierarchy. In your TMS you decide for a which languages to offer. This system will monitor the original language and make a copy under a new language folder. If there are additions or changes to original the TMS will see that and copy those to the translated folders.

The TMS would track what is translated and also approved so you can present a status or offer a threshold how much need to be translated before you publish. i.e. it could be ok if 99% is translated for the remaining 1% it would show the untranslated original text.

This way I think the effort on your side could be kept at minimum. On the publishing side you need to provide a GUI for selecting language version, you would also need to offer some official language tags so systems can chose automatically based on the users "locale". You would need to figure out the integration to the TMS, how the TMS can read and scan your sql storage. And how the TMS can add a new language and write and update the translated language version.

I assume that the images would be the same for all languages versions, so only one assets library to serve all language versions.

With TMS I think Crowdin could be a good start, you already use them for localization of the Bookstack application and they seem to be willing to help out in doing this. I have no relation to Crowdin by the way, I'm just trying to build my toolchain for translating my GUI and Guides. If Bookstack and Crowdin where integrated this would have solved my needs.

**Comment by jasell** [comment]
We are still very much interested in a solution like Bookstack but we require content translation support.

Can we some how help to make this happen?
- Can we contribute or cover some of the cost for the integration.
- We could also assign development resources to help out.

What would be your preferred option to get this moving?

Can Bookstack use GitHub for storage? If so many translation system already have integrations to GitHub.

## Short outline what needs needs to be done...
(I'm not a developer so I might got some things wrong here...)

- Bookstack to Translation management system integration
Would storing texts and images on GitHub be a solution?
  - There need to be a function that upload all new texts to the translation management system.
  - There also need to be a away for the translation system to upload all the translated versions to Bookstack.
- Folder structure would need to change slightly as on the top level there will be language-folders.
  - Where one is created in Bookstack and the others are created and maintained form the the translator system.
  - Parallel with the language folders there need to be a "resource" folder that all language versions will refer to for images etc.
- Bookstack need to make an inventory of available language.
- Bookstack need a language selector.
Either so the user can select themselves or language is specified on the URL when calling bookstack.

**Comment by ssddanbrown** [comment]
Hi @jasell, Thanks for offering support.
Just to quote and re-iterate the statement from my previous comment:

> The primary goal would be multiple languages per page in the current editor experiences. Automation outside of that would probably be delegated to API use via external apps/middleware. Doesn't make sense for us to support specific translation services/platforms.

So I would consider any TMS/external-system-support outside of the scope of this issue. Maybe that could be requested afterwards, once we have the multi-language infrastructure in place, but It's likely I'd still consider that specific solution out of scope of the core project at that time since platform-specific support does not make sense in terms of maintenance and efforts, it's better for us as a platform to focus on providing the API for abstract usage.

> Can we contribute or cover some of the cost for the integration.
> We could also assign development resources to help out.
> What would be your preferred option to get this moving?

To be honest, neither. Someone else developing a feature so wide will likely just guilt me to force focus of my efforts. 
As I've touched upon in previous comments, It's not so much the implementation but also the long term impact of such a feature, relative to the demand. Potentially some commitment of ongoing financial support could help, but I'm not sure how to officially formulate that, I may have to explore similar circumstances in other open source projects.

Since my last comment, I am now aiming to work on the project full time, and that's my current plan, which makes such a feature more viable to proceed with. It's now just a case of why this feature should take priority over any other work, especially for it's impact.

**Comment by jasell** [comment]
@ssddanbrown, Thanks for elaborating on your reason.

So if we decided to make such integration, we still depend on some support on the architecture. As how translated pages are stored. Page by page, book by book or shelf by shelf.

Most likely all language versions will refer to the same resources, etc.

Would you be willing to look into this?

I asked before, would GitHub be a candidate for storing the content during editing? That would simplify integration with TMS system as these often has an integration to GitHub.

**Comment by ssddanbrown** [comment]
@jasell 

> Would you be willing to look into this?

Yes, that's the entire idea of this issue and my comments in regard to working on this are in the last paragraphs of my previous comment.

> I asked before, would GitHub be a candidate for storing the content during editing?

No, As I stated above multiple times I would not look to support specific external platforms in the scope of this issue/feature. You would be free to use GitHub in your own translation pipeline/process (Via connecting with BookStack APIs or extension systems) but our focus would be the core system and the provision of a generic API for any external usage.

**Comment by BashTec** [comment]
Hi together,
first of all many thanks for this great project!
For scientific and research institutes in Germany, the lack of multilingualism is unfortunately an obstacle for widespread use.

We would love to see this feature and I think it could significantly drive the adoption of BookStack.

**Comment by keyhelp** [comment]
Hello!
just came here to thank you for this great software!
I would also highly appreciate the feature discussed here ;).

Right now I'm using different shelves to manage different languages.
That's fine, as long as the user 
- doesn't use the search function, he might then get results in languages he doesn't speak 
- uses the "Books" link in the upper right corner, where he would see every book in every language
- or use the breadcrumbs navigation " > " button

Nice to hear, you may have time in the future to implement such a feature! Keep up the fantastic work ;).

**Comment by angevi** [comment]
Is there an update on this issue? There is strong interest from three different sister-social enterprises here. The need arises from countries with more than one language. Creating knowledge bases in several languages is an issue of social inclusion.

**Comment by jasell** [comment]
We are still open for this and can contribute with either money or time.
I would suggest an integration to Crowdin, as that tool is already tied to BookStack for translating the UI.

**Comment by rmibelgium** [comment]
Same interest on the side of our organization. We're continuing to analyse how we could do that with Logical Theme and without forking the code.

**Comment by Silverlan** [comment]
I would also like to express my interest for this.
Artificial Intelligence has become exceptionally good at translating lately and all that's stopping me from using that to offer my wiki in a bunch of different languages is the lack of this feature.

**Comment by 2-click** [comment]
Honestly that feature would be so great

**Comment by jasell** [comment]
@ssddanbrown would you have time to share an update on this topic?
Seems like we are a several who are interested in this and we can probably contribute if you can put forward what could help you prioritize this request, whether it is funding or development resources?

**Comment by ssddanbrown** [comment]
> would you have time to share an update on this topic?
> Seems like we are a several who are interested in this and we can probably contribute if you can put forward what could help you prioritize this request, whether it is funding or development resources?

The primary blocker for me, as alluded to in my previous comments, is the lack of motivation for what would be taking on a significant amount of technical, maintenance & support dept in the system complexity this would add.

While funding would help that motivation, it would really need to be long-term funding since the costs are long-term. While I've had several offers from businesses in funding this feature now, I've never had a single actual figure put forward, since most seem to act like this is a one-off business transaction that they want a deal on.

While I'm very grateful @jasell that you have offered funding and resources, I've been hesitant to inquire further since within this thread you have seemed to misunderstand the direction and scope I'd want to take for this feature. I have responded above three times that external systems would be out of scope for this core request, and you have still requested that since.

This (The core feature, not external platform support) is still something I want to implement, I just need to find the right moment and motivation. Might not be too far as I look what add next in the roadmap. The initial move will probably be me starting threads, and linking them here, to ask questions on usage and/or put out initial implementation proposals for feedback.

**Comment by jasell** [comment]
Thanks for sharing your update @ssddanbrown .
We are willing to discuss either a one time fee or subscription for this, whatever could motivate you do this.

My need is a support for authoring, translating and rendering of my guides. I see bookstack currently supporting two of these three needs. 

My suggestions on external system was an attempt to reduce the work on you, but that is not a requirement as long we can get the toolchain/pipeline for the three needs mentioned above. To me it seems that building a complete TMS is harder than making a stable integration to an existing TMS. As you already have a TMS integration for your own UI I naturally looked at that as a first option. But if you wish to provide a TMS as part of bookstack I see no problem to use it, but it appears as more work and responsibility for you.

If you are not considering this during foreseeable future, would you consider to draft some guidelines on how an integration to a TMS would work? As you know some about TMS (as you translate your own UI) and you are the one with the best knowledge of Bookstack and its API. 
I could think of to setups, you tell which the better option:
- Direct integration: Bookstack <- integration -> Crowdin
- GitHub as intermediate: Bookstack <-integration -> GitHub <- standard integration -> Crowdin

One benefit of using an external TMS is that we can use it for our own UI as well (we currently use Crowdin as you do), then the translation memory can be shared across the Guides and UI, as I assume your TMS would only support your guides.

**Comment by ssddanbrown** [comment]
Hello everyone :wave: 

I've been doing a fair bit of thinking on this recently.
I was going down the road of changing a lot of our data structure to support language variation, but as I went deeper the added complexities seemed to make less sense and be less worthwhile as they built up, and this led me to think I may be approaching this in the wrong way, thinking about technical implementation first over the fundamental requirements & needs here, and that there may be simpler alternatives to explore that would be better overall.

Therefore, to held explore options, I'd like to approach this at a different angle with the help of feedback from those that desire this feature.
So here's my question to you:

*Right now it's technically possible to service multiple languages through copied/duplicated content, like having separate books for different languages.  Compared to what you need (and by that I really mean only what you need, not what you think might be good or what might be "interesting") how do existing options like that fall short?*

**Please think about your answer only in respect to user possibilities, goals and outcomes rather then thinking about a specific technical implementation you may desire.**

As a pre-emptive reminder: external platforms & translation management systems, outside of our existing standard REST API, are out of scope for this conversation.

**Comment by angevi** [comment]
For a simple book with max two permission levels and max. two languages, it is no problem to have two books. It becomes complicated to manage several books when: 1.) The permissions are more complicated, 2.) The end users switch between languages to better understand, 3.) There are more than two languages. It becomes higher specialized and knowledge bases in organizations need to be easy to use/easy to access/ easy to manage. Especially when employees are diverse in age, background, language.

**Comment by jasell** [comment]
Thanks for looking into this!
I haven’t looked into all you can do with API:s, so I might go out bounds here, sorry about that.

We are documenting our software, Guide is the primary assets, but there is release notes and additional way of working content and other value adding information.

I could guess for every language the system generates a new bookshelf with books in the new language. Our translators will then translate and publish when done.

Our users can select a language through their personal settings and will have read-only guest access.

Authors and translators will have a dedicated login. We don’t need to control which books or languages they can edit/translate, but it would be a bonus if we could.

We have a lot of illustration which we don’t intended to translate, so I would like all translated “copies” to use the same picture library. It should be nice, but no must, to replace some pictures in the translated versions, but that would be an exception.

What I understand since before, your plan is to build a “complete” translation management system, rather than provide a default integration to an existing TMS, like the one you have for the application. That is ambitious, we are happy with either solution.

**Comment by Tagirijus** [comment]
> @ssddanbrown : how do existing options like that fall short?

I hope I understand it correct that you ask, what would fall short, when basically doing some kind of "workaround" for the multi-language-content by e.g. copying a book. A user already mentioned some points, which I can second:

- When using the search, "unwanted" languages might pop up in the results for the user as well.
- Language-books would be listed under `Books` as well. So it would list languages the user might not understand and if they understand it, it would basically "double the content" in the end.
- When having a book-copy for a new language, one would have to cross-reference the languages manually somehow so that the user could "switch" the language if needed.

This comes to my mind offhand.

**Comment by ASparwasser** [comment]
I would really appreciate this feature!!!

**Comment by jessicakuijer** [comment]
me too please!

**Comment by beeing** [comment]
> Hello everyone 👋
> 
> I've been doing a fair bit of thinking on this recently. I was going down the road of changing a lot of our data structure to support language variation, but as I went deeper the added complexities seemed to make less sense and be less worthwhile as they built up, and this led me to think I may be approaching this in the wrong way, thinking about technical implementation first over the fundamental requirements & needs here, and that there may be simpler alternatives to explore that would be better overall.
> 
> Therefore, to held explore options, I'd like to approach this at a different angle with the help of feedback from those that desire this feature. So here's my question to you:
> 
> _Right now it's technically possible to service multiple languages through copied/duplicated content, like having separate books for different languages. Compared to what you need (and by that I really mean only what you need, not what you think might be good or what might be "interesting") how do existing options like that fall short?_
> 
> **Please think about your answer only in respect to user possibilities, goals and outcomes rather then thinking about a specific technical implementation you may desire.**
> 
> As a pre-emptive reminder: external platforms & translation management systems, outside of our existing standard REST API, are out of scope for this conversation.

Hi, just my 2 cents:

The feature needs to address usability for both the authoring / editor and viewer. The technical implementation need not to be complex as most of the time, it is more towards UI.

For authoring side, usually we need to know the last updated date/time of the related pages (same page, diff language). This will help authors to see if some pages need to be updated for different languages. For a simple "diff" UI on the pages, we can just place 2 pages side by side on independent scrollable iframe / divs.

To ease the page configuration (permissions), the rest of the linked pages can be in-synced with the parent page, where all settings to be based on.

As for viewer side, they just need a language selector (can be in header / footer), which will show available languages for that page. Navigation to another page will remember the current language and if the destination does not have the language, it will fall-back to the primary page.

Hope that this is something doable and this project looks awesome!

Have a great day.

**Comment by Man-in-Black** [comment]
I also am very interested in this feature, but I also can understand the concerns about it.
But maybe there is a "simpler" solution to that.
Since automatic translations gotten better and better over the last years (imho), maybe it could be a compromise to use the builtin translation of nearly every browser nowadays.
So each user can translate the content to whatever language he needs.

The only thing which should be added for wider use would be a button somewhere in bookstack which translates the content to the desired language or at least opens up the menu from the browser.

I also know that this is not perfect, but its better than nothing.

**Comment by jasell** [comment]
We tried using Chromes built in translation for both our software and the online guide but the result was far from acceptable, some laughter though. We seen better result when trying some AI-tools.

The challenge lays in the special terminology, invented words and abbreviations, we wish to translate from English to Japanese and Chinese...

**Comment by jeansanson** [comment]
I am also interested in this feature.

**Comment by hieste** [comment]
Hello @all,
Here is a smooth opensource solution: 
https://github.com/LibreTranslate/LibreTranslate

Perhabs it is a possipility to do a post-hook event  and translate the content...

**Comment by jasell** [comment]
> Hello @ALL, Here is a smooth opensource solution: https://github.com/LibreTranslate/LibreTranslate

Have you made any attempts to integrate this with Bookstack?

**Comment by hieste** [comment]
Yes. We install an local instance of libretranslate on a powerfull host. We wrote a programm to get all pages from bookstack step by step via rest api and check if there was an update. If yes, we remove all html tags and put it to libretranslate via rest api (we translate to 3 languages: German, French, Englisch). The response, we put in seperate tags in bookstack with the key of the language. So you can search in the language you want. 
![image](https://github.com/BookStackApp/BookStack/assets/30119314/e1663857-df2b-4a32-a0ad-57ec70719916)

**Comment by LHBL2003** [comment]
I also carefully read through the article from 2017. I also agree that Bookstack needs a multi-language function, as we and many others are working together more and more globally. A good variant would be like Microsoft, where either the content is available in x human-written languages. Or if it was only written in German, there should be an info in the English page that this page was automatically translated from the original language. When searching, the pages should first be displayed in the preferred language and then in other languages.

I even like the option of using the Google translator to start with. However, for reasons I can't yet explain, this is not possible. I have read here that it has already been used. The Chrome translation is neither suggested locally nor in the demo version of Chrome. Does anyone have any idea why this is the case or does it work for others?

Greetings from Germany, where the global English language is not omnipresent.

**Comment by szabeszg** [comment]
> I even like the option of using the Google translator to start with.

That reminded me of this: https://processwire.com/talk/topic/24567-fluency-integrated-deepl-powered-content-translation/

Which is a good example of how others are dealing with issues like this.

**Comment by cgsmith** [comment]
I would like to offer a possible solution. It definitely requires a big change. Here is how I would go about the changes:

- [ ] Configure middleware for localization
- [ ] Make migration to change names to a `json` datatype. Store localizations as JSON for name fields or small fields like description
- [ ] Add `locale` to the page entity.
- [ ] Allow the middleware to route to the proper page entity depending on the locale that is configured. 

Maybe set a setting for a fallback_locale under settings. That way if a route doesn't exist with the locale the fall back will be present.

I think technical debt is fairly limited with a proper test plan and unit testing. I also think automated translations and outside integrations are outside of the scope.

**Comment by quirkiest** [comment]
We are a multi-lingual company based in Japan and Australia.  As I see it the challenge, from an admin perspective, is that maintaining content in separate places (e.g. books) means that the synchronisation becomes very manual and very likely to fall out of sync quickly.  So while a translation solution does not IMO need to be super-complex, it does need to have some way to allow admins to manage the same content in different languages holistically.
One idea that occurred to me was that perhaps this could work well in-page tabs, which would be a feature not exclusively aimed at translation but which could take us a lot of the way there.  If we can manage both our Japanese and English content  in the same document (page) and have some way for users to flick to the version that they can read it would (at least for our company) be a really nice feature.  And I figure tabs are a cool feature to have anyway...

P.S. in the meantime we are using collapsible blocks - not an awful workaround tbh.

**Comment by db-itech** [comment]
Hey everyone, nowadays we maybe should think that it will be better not to maintain multiple languages at all, but rather move the responsibility to systems or the user translating it via LLMs/AI (or corporate AI LLMs) automatically when it is requested and needed.

That way you only have to maintain one original and up-to-date version of your page/book/etc. and have automated translations available in multiple languages when needed. This would also save some time adding a feature and updating/maintaining it while others use AI for that.

**Comment by carlossierra311** [comment]
Hi @ssddanbrown. I'll try to provide some feedback for one of your questions:

> _Right now it's technically possible to service multiple languages through copied/duplicated content, like having separate books for different languages. Compared to what you need (and by that I really mean only what you need, not what you think might be good or what might be "interesting") how do existing options like that fall short?_

The problem we have found with this approach is that when we copy the original book, all hyperlinks on the new book will point to the original book's pages (i.e. to the original language book instead of the translated language book), so besides translating all our text (which is expected), we also need to rework all the hyperlinks (which is unexpected). (By the way, is this the expected behavior, or are we misusing BookStack?)

Best regards!

**Comment by vmario89** [comment]
i was looking for the same feature of localization too. Changing the ui lang is cool, but it does not help for anything if the content itself is not in the language, the user excepts to see. the only scenario would be to have different shelves for differen languages, yes. but if we want to translate content 1:1, this is really not maintainable

btw. my motivation for thinking about all this: doing a documentation with bookstack for a large open source software project for an already existing community, which is spread accross europe.

i was thinking about doing some stuff like hosting 
* one bookstack instance on sth. like en.docs.yourdomain.org
* another boostack instance on de.docs.yourdomain.org
... sharing the same database of users, groups, etc., but have some kind of master database which means we cannot create or modify new users or groups in all other databases
... with an environment which is fixed to the certain language, so ui language = content language
... synchronizing page ids and other table related stuff, but storing the different language content per page in the certain database

caveats would be:
* many. i think this is highly complex :-(

**Comment by FreeSynergy** [comment]
Hello to everybody and espacially @ssddanbrown 

I know, that this could be a little complex, to do, but as you see, there are not less people, who want this feature. 
I, personally, don't like PHP and I wanted to get rid of it. But I choose Bookstack, because the idea and the features are really great. There are 2 things I'm missing:
- Permissions a little more detailed (for example: Author - can only save as drafts - , Editors - can save drafts and publish them -, Publishers - can only publish content).
- The other is the multilanguage …

What can be the solution?
At first, we can make a variable, witch language it is. This will be stored in the url, after the domain. E.G. bookstacks.org/en or bookstacks.org/en_US … this depends on you.
As now, every Object (Shelves, Books, Chapters, Pages, Images and so on) has an ID - you need to store additionaly the Language_Code AND the Original_ID. (you will store always as foreignkeys the Oringinal_ID when you connect the shelves, the Books and so on together).
The Primary Key is the ID,
The Original ID and Language Code are Unique …

With that Chance, you won't have to rewrite everything … but you can use this, to add a feature, that that can easily expand …
For example:
If you don't want to change much things … you can show every book, ervery caption, every page and so on by making a select WHERE the  ID is … 
If you like to extant that, you can make some great charts, witch language is used more, bind up an KI to translate it automaticly and so on …

I don't think, that this will be easy … but I think, that this is very usefull in our nower days …

Thanks for reading. And maybe we will have a good announcement …

Have a good before Christmas time …
Kal

**Comment by BartoszStepien-3DArtist** [comment]
Hi @ssddanbrown. First of all, thank you for the software. It is great. I join the request for multi-language support. +1

**Comment by UnSciPreneur** [comment]
+1 on the support of multi-language content.

**Comment by BenjaminKobjolke** [comment]
+1

**Comment by Superjamp** [comment]
+1

**Comment by ssddanbrown** [comment]
Please don't add any +1 style comments, they add noise and notification, so I will delete them. Just add a 👍 reaction to the original post instead.

---

Just an update on this request, I have recently spent a while thinking through options again.
The more I think about it, the more it makes sense that this is managed via multiple instances instead of at the same instance, since that allows a lot of flexibility without adding a lot of the BookStack level complexity required. 
Of course that may be more burdensome for smaller scenarios, but it's tricky to judge where on the scale the average scale of desire for this sits relative to that burden. 
Regardless, a multi-instance setup would still require a fair bit of built in helper logic & UI to solve many of the challenges noted here in comments; Things like language switching control, cross-lang content sync/comparison/referencing etc...
Then there could be extra challenges of a multi-instance setup, upon the maintenance/management of multiple instances. Things like permission management and user management (which could be partially helped via SSO).

A tricky factor is that the best approach can change quite a bit depending on needs, user-approach/desire and scale, hence I tend towards a more flexible yet higher burden implementation.

I've kinda burnt myself out thinking about this for now, but will try to ruminate upon this potential approach/direction.

**Comment by jasell** [comment]
Thanks for the update!
Can you share anything about the timing, to manage our expectations?
Does this mean it's on the roadmap and will happen within a year or is this a next millennium thing?

**Comment by ssddanbrown** [comment]
@jasell Can't promise anything on timings. History would suggest that it won't be this year.
I usually cycle through ideas & requests, then rapidly proceed if I see a viable/suitable implementation path forward, hence no longer term roadmap. I checked in on this as a highly requested feature, to potentially implement for the 10 year anniversary of BookStack, but burnt out on it while planning, so will just have to cycle back in the future, probably after thinking though the path described above in the background. That might be later this year, or might not.

**Comment by Superjamp** [comment]
thanks, too :)

I understand that adding this feature after the fact might be really tricky. I haven't looked at the code in detail, but I think you know best how much time and headaches it will take to do so.

It is also definitely a challenging task for content creators to implement books in a [n] instance approach and for the organization to keep multiple instances with roles and permissions on par. Synchronization between instances could probably be achieved with a few scripts. An easy workaround for content creators today would be to structure their books using tags and/or using tags in their titles. (but messy)

From a user perspective, I'd like to have the content when it's available in my browser or preferred language. Then the content creator has to update both, and that's where it starts to fall apart if they forget to do so. Maybe there can be a feature implemented to allow the content creator to translate / update the localized page in a desired language with DeepL for example.
But now everyone knows that most browsers will try to translate the page for you... except for media content. And it is highly questionable whether localized content in this Project is really a necessary feature, given the current effort required to implement it. 

In the context of a university using BookStack, it's a very necessary feature especially for international students.
And we (TU Braunschweig) are very interested in this feature request to get a good working implementation in BookStack.

Personally, I don't mind having only one supported language, it's perfectly fine for the scope of this project.

**Comment by jasell** [comment]
> But now everyone knows that most browsers will try to translate the page for you... except for media content

We tried this, but at least google's translation is far from good. We use bookstack as a guide for our software. Our software is highly specialized with a rather unique terminology, hence the need for the guide. 

If google would offer a way to suggest corrections to its translation it might solve the problem and also improve the translation capability over time, but google doesn't want to allow this for some reason.

**Comment by quirkiest** [comment]
Multiple instances sounds complex and disconnected, to be honest.  Perhaps the result could be achieved with a much more modest UI component change, similar to the existing accordion.  If we could include a tab-type control which allowed formatted content inside then you could store each language's content for that section/page etc inside an individual tab.  The tab UI feature would possibly be handy for a lot of other things, too.

**Comment by Hexodus** [comment]
I'm a bit late to this party but anyways. My first thought was: 
Careful with language features! They can cause complexity to explode and introduce rigid, fragile structures that only a few users actually need – and that end up being a pain in the ass for the rest of us.

IMO, Bookstack is great because of its flexibility, so whatever we do, the solution should be dead simple. Something that opens up new possibilities without requiring deep changes. That means no full-blown localization or translation, AI or not – those are too specific and hard to implement.

Instead, **I’d use the content tag system to mark the language of content**. This should be inherited down the hierarchy – so a book tag like locale: en would also apply to chapters and pages, but could still be overridden at any level. Then just stick a flag icon next to it for visual indication, add a clean language selection dropdown, and that's it. 

That alone would be super useful for me. I wrote an extension that lets me drag & drop citations from Zotero, and I need a way to determine the language so I can adjust the link content accordingly. Right now, there's no good way to do that since tags aren’t inherited.

**Comment by FreeSynergy** [comment]
> Careful with language features! They can cause complexity to explode and introduce rigid, fragile structures that only a few users actually need – and that end up being a pain in the ass for the rest of us.

I've an other opinion …
Language features are very helpful to bring people together, that have an other language. There are always workarounds and this may be for you a good idea, to use it. But if you want more, than only to supply your "people", than a good implemented feature like "language" can be helpful for many people.

In modern "wikis", this feature is standard … the question is, why should we don't implement this and make this "wiki" to a better version? Not only technically, but also socially …

If you are implementing this in a wrong way, you're right. But if you do ist right, you don't have too much trouble with it.

First thing is, because you mentioned it:
- You can use an internal "tag" for language. 

If this ist implemented, than the biggest thing is done (except, when there are some workaround, which are not standard …)

Next thing is to implement dates-, timeformat and so on … but this is second, third … steps …

First step is to save the language tag with every data you store …

**Comment by add-n2x** [comment]
I've successfully implemented BookStack now within two non-profit organizations. One of them, our community radio station, has some big need for multi-language support. 

Is there an update on the roadmap for this feature? We offer you to send you a free t-shirt, if this feature could be implemented at some point. 😸 

Thanks for all of your great work, Dan!

**Comment by aliozgur** [comment]
Not an ideal solution, but below you can find a script which utilizes [Translator and Language Detector APIs](https://developer.mozilla.org/en-US/docs/Web/API/Translator_and_Language_Detector_APIs) 

The script;

* Places a Translate dropdown on the top navigation menu
* Translates main.content-wrap, .book-tree, .sidebar-page-nav, .entity-list content on the fly
* Displays translation progress div
* If api's are not supported displays error div
* Defaults to html lang attribute value
* Saves users translation pref in a cookie

**NOTES:** 
* Works with Chrome version 141+


<details><summary>Code</summary>
<p>

```javascript
<script>
document.addEventListener("DOMContentLoaded", () => {
  // ===================== CONFIG =====================
  const OFFERED_LANGS = ["en", "de", "fr", "es", "it", "ru", "ar"]; // allowed targets
  const COOKIE_NAME = "pref_lang";          // "default" => do not auto-translate
  const COOKIE_DAYS = 180;
  const MODEL_TIMEOUT_MS = 45000;           // model download timeout
  const SECTION_CONCURRENCY = 16;           // parallel translations per section

  const LINKS_DIV = document.querySelector("header nav .links");
  // Translate sections concurrently: content, book tree(s), sidebar(s), entity list(s)
  const SCOPES = [
    { el: document.querySelector("main.content-wrap"), label: "content" },
    ...Array.from(document.querySelectorAll(".book-tree")).map(el => ({ el, label: "book tree" })),
    ...Array.from(document.querySelectorAll(".sidebar-page-nav")).map(el => ({ el, label: "sidebar" })),
    ...Array.from(document.querySelectorAll(".entity-list")).map(el => ({ el, label: "entity list" })),
  ].filter(s => !!s.el);

  if (!LINKS_DIV || SCOPES.length === 0) return;

  // ===================== Cookie utils =====================
  function setCookie(name, value, days) {
    const d = new Date(); d.setTime(d.getTime() + days*24*60*60*1000);
    document.cookie = `${name}=${encodeURIComponent(value)}; expires=${d.toUTCString()}; path=/; SameSite=Lax`;
  }
  function getCookie(name) {
    const m = document.cookie.match(new RegExp("(^| )"+name+"=([^;]+)"));
    return m ? decodeURIComponent(m[2]) : null;
  }
  function normalizeLang(l) { return (l || "").toLowerCase().split("-")[0]; }

  // ===================== Status bar (in-flow, persistent) =====================
  let bar, barInner, barText, barBtn, barPct, barSections;
  const sectionChips = new Map(); // label -> <span>

  function ensureBar() {
    if (bar) return;
    bar = document.createElement("div");
    bar.style.position = "relative";
    bar.style.background = "rgba(34,34,34,.92)";
    bar.style.color = "#fff";
    bar.style.fontSize = "13px";
    bar.style.padding = "8px 12px";
    bar.style.border = "1px solid rgba(255,255,255,.08)";
    bar.style.borderRadius = "6px";
    bar.style.marginBottom = "10px";
    bar.style.width = "100%";
    bar.style.boxSizing = "border-box";
    bar.style.pointerEvents = "none"; // wrapper never blocks clicks

    barInner = document.createElement("div");
    barInner.style.display = "flex";
    barInner.style.gap = "10px";
    barInner.style.alignItems = "center";
    barInner.style.pointerEvents = "auto";     // buttons are clickable
    bar.appendChild(barInner);

    barText = document.createElement("span");
    barText.textContent = "Preparing translation…";

    barBtn = document.createElement("button");
    barBtn.type = "button";
    barBtn.textContent = "Start";
    barBtn.style.padding = "4px 10px";
    barBtn.style.cursor = "pointer";
    barBtn.style.border = "1px solid rgba(255,255,255,.25)";
    barBtn.style.background = "transparent";
    barBtn.style.color = "#fff";
    barBtn.style.borderRadius = "6px";
    barBtn.style.display = "none";

    barPct = document.createElement("span");
    barPct.style.marginLeft = "auto";
    barPct.textContent = "";

    barInner.appendChild(barText);
    barInner.appendChild(barBtn);
    barInner.appendChild(barPct);

    // Per-section chips row
    barSections = document.createElement("div");
    barSections.style.display = "flex";
    barSections.style.flexWrap = "wrap";
    barSections.style.gap = "6px";
    barSections.style.marginTop = "6px";
    barSections.style.opacity = "0.9";
    bar.appendChild(barSections);

    const attach = SCOPES[0]?.el || document.body;
    attach.insertBefore(bar, attach.firstChild);
  }
  function barStatus(text) { ensureBar(); barText.textContent = text || ""; barText.style.color = "#fff"; bar.style.background = "rgba(34,34,34,.92)"; }
  function barError(text)  { ensureBar(); barText.textContent = text || "Translation error"; barText.style.color = "#ff6b6b"; bar.style.background = "rgba(34,0,0,.25)"; barPct.textContent = ""; barBtn.style.display = "none"; }
  function barProgress(done, total) { ensureBar(); if (!total) { barPct.textContent = ""; return; } barPct.textContent = `${Math.round((done/total)*100)}%`; }
  function barShowStart(handler) { ensureBar(); barBtn.style.display = "inline-block"; barBtn.onclick = handler; }
  function barHideStart() { if (!barBtn) return; barBtn.style.display = "none"; barBtn.onclick = null; }
  function barDone(msg) { barStatus(msg || "Translated"); /* keep bar in place */ }

  function ensureChip(label) {
    if (sectionChips.has(label)) return sectionChips.get(label);
    const chip = document.createElement("span");
    chip.style.background = "rgba(255,255,255,.08)";
    chip.style.border = "1px solid rgba(255,255,255,.12)";
    chip.style.borderRadius = "999px";
    chip.style.padding = "2px 8px";
    chip.style.fontSize = "12px";
    chip.textContent = `${label}: 0%`;
    barSections.appendChild(chip);
    sectionChips.set(label, chip);
    return chip;
  }
  function updateChip(label, current, total, done) {
    const chip = ensureChip(label);
    if (!total) { chip.textContent = `${label}: 0%`; return; }
    const pct = Math.round((current / total) * 100);
    chip.textContent = `${label}: ${pct}%`;
    if (current === total) {
      chip.textContent = `${label}: ✓`;
      chip.style.background = "rgba(0,128,0,.15)";
      chip.style.borderColor = "rgba(0,128,0,.35)";
    }
  }

  // ===================== Dropdown (matches your site) =====================
  const dd = document.createElement("div");
  dd.className = "dropdown-container";
  dd.setAttribute("component", "dropdown");
  dd.setAttribute("option:dropdown:bubble-escapes", "true");

  const toggle = document.createElement("span");
  toggle.className = "user-name py-s hide-under-l";
  toggle.setAttribute("refs", "dropdown@toggle");
  toggle.setAttribute("aria-haspopup", "true");
  toggle.setAttribute("aria-expanded", "false");
  toggle.setAttribute("tabindex", "0");
  toggle.innerHTML = `
    Translate
    <svg class="svg-icon" data-icon="caret-down" role="presentation" viewBox="0 0 24 24">
      <path d="M7 10l5 5 5-5z"></path>
      <path d="M0 0h24v24H0z" fill="none"></path>
    </svg>
  `;
  const menu = document.createElement("ul");
  menu.className = "dropdown-menu";
  menu.setAttribute("refs", "dropdown@menu");
  menu.setAttribute("role", "menu");
  menu.style.display = "none";

  const LABELS = {
    default: "Default",
    en: "English",
    de: "German",
    fr: "French",
    es: "Spanish",
    it: "Italian",
    ru: "Russian",
    ar: "Arabic"
  };
  const items = ["default", ...OFFERED_LANGS];
  items.forEach((code) => {
    const li = document.createElement("li");
    const a = document.createElement("a");
    a.href = "#";
    a.className = "icon-item";
    a.textContent = LABELS[code] || code.toUpperCase();
    a.addEventListener("click", (e) => {
      e.preventDefault();
      menu.style.display = "none";
      toggle.setAttribute("aria-expanded", "false");
      setCookie(COOKIE_NAME, code, COOKIE_DAYS);
      location.reload(); // always from pristine content (or do nothing if default)
    });
    li.appendChild(a);
    menu.appendChild(li);
  });

  dd.appendChild(toggle); dd.appendChild(menu); LINKS_DIV.appendChild(dd);
  toggle.addEventListener("click", () => {
    const open = menu.style.display === "block";
    menu.style.display = open ? "none" : "block";
    toggle.setAttribute("aria-expanded", String(!open));
  });
  document.addEventListener("click", (e) => {
    if (!dd.contains(e.target)) { menu.style.display = "none"; toggle.setAttribute("aria-expanded", "false"); }
  });

  // ===================== Helpers for multi-scope =====================
  function getSampleText(max = 1500) {
    const parts = [];
    for (const s of SCOPES) {
      const txt = (s.el?.innerText || "").trim();
      if (txt) parts.push(txt);
      if (parts.join(" ").length >= max) break;
    }
    return parts.join(" ").slice(0, max);
  }

  function collectTextNodesFor(el) {
    const SKIP = new Set(["SCRIPT", "STYLE", "NOSCRIPT"]);
    const arr = [];
    if (!el) return arr;
    const walker = document.createTreeWalker(el, NodeFilter.SHOW_TEXT, {
      acceptNode(node) {
        const p = node.parentElement;
        if (!node.nodeValue || !node.nodeValue.trim()) return NodeFilter.FILTER_REJECT;
        if (!p || SKIP.has(p.tagName) || p.isContentEditable) return NodeFilter.FILTER_REJECT;
        if (p.closest('form')) return NodeFilter.FILTER_REJECT; // skip any form content
        if (getComputedStyle(p).display === "none" || p.closest?.("[data-no-translate]")) return NodeFilter.FILTER_REJECT;
        return NodeFilter.FILTER_ACCEPT;
      }
    });
    let n; while ((n = walker.nextNode())) arr.push(n);
    return arr;
  }

  // ===================== Translation Core =====================
  let translator = null, detector = null;

  async function withTimeout(promise, ms, label) {
    let t; const timeout = new Promise((_, rej) => t = setTimeout(() => rej(new Error((label||"Operation")+" timed out")), ms));
    try { return await Promise.race([promise, timeout]); } finally { clearTimeout(t); }
  }

  async function detectorAvailability() {
    try {
      if (!("LanguageDetector" in window)) return "unavailable";
      if (LanguageDetector.availability) return await LanguageDetector.availability({});
    } catch {}
    return "available";
  }

  async function detectSourceLanguage(invokedByUser) {
    // Try detector. If not available (or needs gesture without one), fall back to "tr".
    const avail = await detectorAvailability();
    const canUseNow =
      avail === "available" ||
      ((avail === "downloadable" || avail === "downloading") && invokedByUser);

    if (!canUseNow) return "tr";

    try {
      detector = detector || await LanguageDetector.create({});
      const sample = getSampleText(1500);
      const results = await detector.detect(sample);
      return results?.[0]?.detectedLanguage || "tr";
    } catch {
      return "tr";
    }
  }

  async function translatorAvailability(sourceLang, targetLang) {
    try {
      if (Translator.availability) {
        return await Translator.availability({ sourceLanguage: sourceLang, targetLanguage: targetLang });
      }
    } catch {}
    return "available";
  }

  async function createTranslator(targetLang, invokedByUser) {
    const rawSource = await detectSourceLanguage(invokedByUser);
    const sourceLang = normalizeLang(rawSource);
    const normTarget = normalizeLang(targetLang);

    if (sourceLang === normTarget) return { translator: null, reason: "same_language", sourceLang };

    const avail = await translatorAvailability(sourceLang, normTarget);
    if (avail === "no" || avail === "unsupported") return { translator: null, reason: "unsupported", sourceLang };

    if (avail === "downloadable" || avail === "downloading") barStatus("Downloading language model…");
    else barStatus("Preparing translator…");

    const t = await withTimeout(
      Translator.create({ sourceLanguage: sourceLang, targetLanguage: normTarget }),
      MODEL_TIMEOUT_MS,
      "Model download"
    );
    return { translator: t, reason: null, sourceLang };
  }

  // Translate a single section with limited concurrency
  async function translateSection(label, nodes) {
    ensureChip(label);
    const total = nodes.length;
    let completed = 0;

    // Worker loop grabbing the next index atomically
    let index = 0;
    async function worker() {
      while (true) {
        const i = index++;
        if (i >= total) return;
        const node = nodes[i];
        try {
          const out = await translator.translate(node.nodeValue);
          if (typeof out === "string" && out) node.nodeValue = out;
        } catch (e) {
          // no-op, skip failed node
        } finally {
          completed++;
          sectionProgressTick(label, completed, total);
        }
      }
    }

    // Kick off up to SECTION_CONCURRENCY workers
    const k = Math.min(SECTION_CONCURRENCY, total || 1);
    await Promise.allSettled(Array.from({ length: k }, () => worker()));
  }

  // Global progress aggregator
  let globalTotal = 0, globalDone = 0;
  function sectionProgressTick(label, completed, total) {
    globalDone++;
    updateChip(label, completed, total, globalDone);
    barProgress(globalDone, globalTotal);
  }

  async function translateNodesConcurrent() {
    // Prepare sections
    const sections = SCOPES.map(s => {
      const nodes = collectTextNodesFor(s.el);
      ensureChip(s.label);
      updateChip(s.label, 0, nodes.length, 0);
      return { label: s.label, nodes };
    });

    globalTotal = sections.reduce((sum, s) => sum + s.nodes.length, 0);
    globalDone = 0;

    if (!globalTotal) { barDone("Nothing to translate"); return; }

    barStatus("Translating…");

    // Run all sections concurrently
    await Promise.allSettled(
      sections.map(s => translateSection(s.label, s.nodes))
    );

    barDone("Translated");
  }

  async function translateFlow(targetLang, invokedByUser) {
    try {
      if (!window.isSecureContext) { barError("HTTPS required for on-device translation."); return; }
      if (!("Translator" in window)) { barError("Translator API not available in this browser."); return; }

      // Permissions-Policy quick check
      try {
        const pp = document.permissionsPolicy || document.featurePolicy;
        if (pp?.allowsFeature && !pp.allowsFeature("translator")) { barError("Blocked by Permissions-Policy: translator."); return; }
      } catch {}

      ensureBar();
      barStatus("Preparing translation…");

      // Probe: if model needs a gesture and we're auto-running, show Start
      const srcProbe = normalizeLang(await detectSourceLanguage(invokedByUser));
      const probeAvail = await translatorAvailability(srcProbe, normalizeLang(targetLang));
      if (!invokedByUser && (probeAvail === "downloadable" || probeAvail === "downloading")) {
        barStatus("Click Start to load translation model…");
        barShowStart(async () => { barHideStart(); await translateFlow(targetLang, true); });
        return;
      }

      const result = await createTranslator(targetLang, invokedByUser);
      if (!result.translator) {
        if (result.reason === "same_language") { barDone("Already in selected language"); return; }
        if (result.reason === "unsupported") { barError(`Unsupported language pair (${result.sourceLang} › ${normalizeLang(targetLang)}).`); return; }
        return;
      }
      translator = result.translator;
      await translateNodesConcurrent();
    } catch (err) {
      if (String(err?.name) === "NotAllowedError") {
        barStatus("Click Start to begin translation…");
        barShowStart(async () => { barHideStart(); await translateFlow(targetLang, true); });
        return;
      }
      barError(err?.message || "Translation failed.");
      console.error(err);
    }
  }

  // Mark specific forms as no-translate (just in case)
  (function markFormsNoTranslate() {
    const EXCLUDE_SELECTORS = [
      'form[refs="list-sort-control@form"]',
      'form[action*="/preferences/change-sort/shelf_books"]'
    ];
    EXCLUDE_SELECTORS.forEach(sel => {
      document.querySelectorAll(sel).forEach(el => el.setAttribute('data-no-translate', ''));
    });
  })();

  // ===================== Auto-translate on load =====================
  (async function maybeAutoTranslateOnLoad() {
    const pref = getCookie(COOKIE_NAME); // "default" or a lang code
    if (pref === "default") return;

    let target = null;
    if (pref && pref !== "default") {
      target = normalizeLang(pref);
    } else {
      const htmlLang = normalizeLang(document.documentElement.getAttribute("lang") || "");
      if (htmlLang && htmlLang !== "tr" && OFFERED_LANGS.includes(htmlLang)) target = htmlLang;
    }
    if (!target) return;

    ensureBar();
    barStatus("Preparing translation…");
    await translateFlow(target, /*invokedByUser*/ false);
  })();
});
</script>

```

</p>
</details>

**Comment by Transer1** [comment]
> If you have not planned this functionality yet, I would like to collaborate with you by implementing it.

Hey, i want to underline the want for multilingual support.
OP, you still down to collab?

**Comment by O5ten** [comment]
This feels like it could be an AI integration feature. I just want a simple way to translate pages "automatically" if they contain a specific tag and connect it to an LLM of my choice that updates the content of the english version.

**Comment by axcro7** [comment]
The thing with AI content is that it still takes a human to check the output before publishing. 
We want to provide detailed instructions for end users all over the world. We would only have like three languages but absolutely need the translations to be exact. 

That being said, I could think of two workflows:
A. Write the docs three times and place links to the translations on every page. This might be the easiest to implement but would involve a lot of manual work by the authors and introduce human errors. 
B. Have a single page with textblocks tagged with a language. The user would then have a language selection somewhere on the top (drop-down menu / flag icons?) and only the textblocks with the right language tag would be displayed. No idea how one would implement this ..

**Comment by ssddanbrown** [comment]
BookStack is now [managed on Codeberg](https://codeberg.org/bookstack/bookstack).

As part of this change, we're closing all existing GitHub issues and pull requests.
You'll instead find this issue on Codeberg here: 
https://codeberg.org/bookstack/bookstack/issues/569
  
Any further discussion for this will now be done on Codeberg.
Therefore this GitHub thread will be locked.
